### ISSUE TITLE
Clearcasting for Mages and Shamans is bugged.

### ISSUE THREAD

**Issue by Velite12** [slowtorta/turtlewow-bug-tracker#376]
**Please, provide a clear description what the bug is:**

Clearcasting for Mages and Shamans is not blizzlike:

So TLDR, Blizzlike: 
Mage clearcasting: procs on land, consumed on cast.
Shaman clearcasting: procs on cast, consumed on cast.
Vmangos:
Mage clearcasting: procs on land, consumed on land (visually but in reality spell is used)
Shaman clearcasting: procs on land, consumed on land (visually but in reality spell is used)

**Steps to reproduce the behavior with as much detail as possible:**

1. Cast a Frostbolt, with Arcane Concentration talented as a mage and you will get clearcasting on land. Then cast another frostbolt, the clearcast won't be consumed until you land.
2. Cast a Lightning Bolt with Elemental Focus talented as a shaman, you will not get clearcasting until the Bolt actually hits, and it won't be consumed on the next cast until the bolt hits.
3. In both cases, Casting an instant spell while the 2nd bolt is in transit will cost mana, which means that the Clearcasting is consumed in reality but not visually.

**IDs of affected creatures, items, quests or spells with a link to the relevant page:**

Spell:

Mage Arcane Concentration and Clearcasting:
https://database.turtle-wow.org/?spell=12577
https://database.turtle-wow.org/?spell=12536

Shaman Elemental Focus and Clearcasting:
https://database.turtle-wow.org/?spell=16164
https://database.turtle-wow.org/?spell=16246

**Expected behavior. Describe how it should work:**

Mage clearcasting: procs on land, consumed on cast.
Shaman clearcasting: procs on cast, consumed on cast.

The following videos are from Classic WoW vanilla and Retail Vanilla

https://youtu.be/bMWKNeIISmo?t=104

https://www.youtube.com/watch?v=DXMlgQVKxMA&t=122s

https://youtu.be/JE_KzAEBIVc?t=48

https://www.youtube.com/watch?t=53&v=pwCW9JiYwqY&feature=youtu.be

https://youtu.be/NEujh_Dcheg?t=710

![image](https://user-images.githubusercontent.com/45871932/137271021-e238e8d8-8008-4903-ac40-3cc56de40d61.png)

You can clearly see that Clearcasting procs before Elemental Mastery is consumed, which means that it did not trigger from the spell landing like Elemental Mastery does.

**Comment by Velite12** [comment]
https://github.com/vmangos/core/commit/cc6545f8fb72852c5c5d99e838e96a9cd91c2a28 has been made to address the issue of shaman clearcasting

**Comment by Nolin7777** [comment]
Possible fix has been applied, please test after next maintenance.
